### ISSUE TITLE
Allow `brace-style` rule to enforce different styles on different block types

### ISSUE THREAD

**Issue by jrsearles** [eslint/eslint#6828]
**What version of ESLint are you using?**
v3.1.0

**What rule do you want to change?**
brace-style

I would like to enable different brace handling based on the type of block. For example, different handling for IfStatements vs FunctionDeclaration (and even FunctionExpression).

Ideally you would be able to use individual options to mimic the predefined formats (

``` json
{
  "brace-style": ["error", { "IfStatement": "new-line", "FunctionDeclaration": "new-line", "FunctionExpression": "same-line" }, { "noCuddledElse": true }]
}
```

BlocksSupported:
`FunctionDeclaration`
`FunctionExpression`
`ArrowFunctionExpression`
`IfStatement`
`TryStatement`
`DoWhileStatement`
`WhileStatement`
`WithStatement`
`ForStatement`
`ForInStatement`
`ForOfStatement`
`SwitchStatement`

Options:
`new-line` - brace always starts on a new line
`same-line` - brace always starts on same line
`ignore` - rule does not apply

Additional options to support "stroustrup" style:
`noCuddledElse` - else keywords start on own line
`noCuddledCatchFinally` - catch/finally keywords start on own line

**What code should be flagged as incorrect with this change?**

With this enabled, the following would be an error:

``` js
function doSomething() 
{
  // function body
}

if (true) {
  // body
} else {
}

invokeCallback(function()
{
});
```

**What happens when the rule is applied to this code now?**

The brace variations do match the flexibility defined. I feel like i'm fighting with my tooling (VSCode) when using the brace-style rule and would like the possibility of adding additional properties for the rule to better fit within our preferred code formatting and development tooling.

**Comment by platinumazure** [comment]
@jrsearles Interesting. Right now `brace-style` accepts only one of three pre-configured brace styles, whereas it sounds like you might want more flexibility at a block vs function level (and whatever other constructs might apply here). In other words, it seems as though you want to just configure your own brace style rules.

I like the concept, but we're going to need to flesh this out a bit further. As a first step, can you please edit your initial post to follow our [proposed rule change template](http://eslint.org/docs/developer-guide/contributing/rule-changes#proposing-a-rule-change) as closely as you can? Thanks!

**Comment by eslintbot** [comment]
Hi @jrsearles, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by jrsearles** [comment]
Thanks - i've updated the comment to comply with the guidelines.

Yes, i would like the rule to be more flexible. I think what would be ideal is if the rule was flexible enough that the predefined formats could be wrappers around individual options.

**Comment by jrsearles** [comment]
I created a plugin (https://github.com/jrsearles/eslint-plugin-brace-rules) which is a reworking of the `brace-style` rule with the flexibility i'm looking for. It's completely backwards compatible (at least all tests pass). I also made it fixable, since this is only stylistic issue it's nice to just let the tool fix it.

I'm more than happy to submit this as a PR. The change i made allowed the second argument to be an object literal with the node types specified with an enum ("always", "never", "ignore") to determine how to enforce the brace rule. Then added a couple additional properties to account for the variance in the Stroustrup style.

BTW, i gotta say the documentation on your site is fabulous. It made creating a plugin a breeze.

If you don't want to incorporate these changes in core feel free to close this issue.

**Comment by platinumazure** [comment]
@jrsearles Wow, nice work on that plugin. I'd love to see this considered for core. I will take the plugin as a proposal.

@eslint/eslint-team I'll champion this. Any endorsements?

**Comment by nzakas** [comment]
I'd like to see some better option names, but otherwise 👍

**Comment by kaicataldo** [comment]
I love this idea! :+1: for me

**Comment by jrsearles** [comment]
Great - what option names are a concern? The enums ("always", "never", "ignore")? Or "noCuddledElse" and "noCuddledCatchFinally"?

**Comment by platinumazure** [comment]
Do we want to consider auto-fix as part of this enhancement request? Haven't seen any discussion on it here.

**Comment by platinumazure** [comment]
@nzakas Can you please clarify which option names you would like to see improved, per your comment [here](https://github.com/eslint/eslint/issues/6828#issuecomment-243847759) and @jrsearles' question [here](https://github.com/eslint/eslint/issues/6828#issuecomment-244620869)? Thanks!

**Comment by kaicataldo** [comment]
One more friendly ping for @nzakas

**Comment by nzakas** [comment]
Sorry, I get so many notifications that pinging me with mentions isn't really going to get my attention. It's best to try to catch me in the chatroom for faster turnaround.

I don't think `newLineFunction` and `newLineBlock` clearly describe what they do when enabled. In fact, I had to re-read this issue five times before I remembered exactly what this proposal was about, which is a good indication that the option names don't say what they do.

Currently, the rule has one named option `allowSingleLine`, which clearly says what it does. I'd like the new options to also clearly say what it does. Is it allowing something? Enforcing something? Disallowing something? Other? And it's unclear how those options relate to the specified brace style ("1tbs", for example).

I *think* the request is to be able to have Allman for blocks and 1tbs for functions, so maybe it would be better to allow a configuration like this:

```
brace-style: [ "error", {
    functions: "1tbs",
    blocks: "allman"
}
```

**Comment by jrsearles** [comment]
Hi @nzakas - thanks for the input. The PR ended up diverging from my original request to give a bit more flexibility so what you were referencing wasn't relevant. Sorry for the confusion. I've updated the original request to match what the PR contains.

To summarize, it enables individual handling of brace-styles per block.

I don't think the predefined styles make sense when it comes to individual blocks, hence the enum. (Though i suppose it might be clearer for the enum to be something line "same-line", "new-line", "ignore")

This ultimately allows you to apply overrides to built in styles.

**Comment by kaicataldo** [comment]
Looks like the original proposal has changed since this was initially accepted and the issue has stalled. We have an open PR for this, so would love to get the discussion going again. I'm still :+1: for this (put a reaction on the original comment as well).

**Comment by platinumazure** [comment]
I don't super enjoy using "1tbs" vs "allman" etc. as values for the individual configurations, because to me those are brace styles that apply equally to all blocks; however, I respect that there is practical value in using values that users will already be familiar with.

The only potential confusion I can see is the use of "allman" vs "stroustrup" for non-if blocks (e.g., function bodies). In that case, they mean the same thing since there is no "else block" to consider. Is that potentially ambiguous, or can we write the rule to respect both of those values?

I think I ultimately favor @jrsearles' [most recent suggestion](https://github.com/eslint/eslint/issues/6828#issuecomment-246882124), which is "same-line", "new-line", and "ignore" (though we would still need to figure out the best way to handle "else" blocks for IfStatement nodes).

I'd really love to see this get in-- let's figure out a good design and get this in pronto! :smile:

**Comment by nzakas** [comment]
Can someone put together a formal proposal using the [template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md). I feel like right now we are just voting on a ghost. :)

**Comment by jrsearles** [comment]
**Please describe what the rule should do:**
I would like to enhance the `brace-style` rule by allowing for different brace handling based on the type of block. For example, different handling for IfStatements vs FunctionDeclaration or FunctionExpression.

Ideally you would be able to use individual options to mimic the predefined formats ("1tbs", "strousoup", "allmans"), providing full backwards compatibility.

**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

WARNINGS
```js
/* eslint brace-style { FunctionDeclaration: "new-line", FunctionExpression: "same-line", IfStatement: "same-line", noCuddledElse: true } */
function doSomething() {
  // function body
}

invokeCallback(function() 
{
});

if (true) 
{
  // body
} else {
}
```

EXPECTED
```js
/* eslint brace-style { FunctionDeclaration: "new-line", FunctionExpression: "same-line", IfStatement: "same-line", noCuddledElse: true } */
function doSomething()
{
  // function body
}

invokeCallback(function() {
});

if (true) {
  // body
} 
else {
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
This is adding flexibility to an existing rule. I feel like i'm fighting with my tooling (VSCode) when using the brace-style rule and would like the possibility of adding additional properties for the rule to better fit within our preferred code formatting and development tooling.

**Comment by platinumazure** [comment]
@eslint/eslint-team Can we please review [this proposal](https://github.com/eslint/eslint/issues/6828#issuecomment-254691834)? I've volunteered to champion. Remember that we can accept the issue even if we believe bikeshedding needs to occur as well. Thanks!

**Comment by not-an-aardvark** [comment]
I'm :+1: for adding block-specific options. However, is there a reason to add new option names such as "new-line" and "same-line" rather than using the existing option names?

I'm imagining a schema like:

```json
{
  "rules": {
    "brace-style": ["error", {"FunctionDeclaration": "1tbs", "FunctionExpression": "allman"}]
  }
}
```

**Comment by platinumazure** [comment]
@not-an-aardvark One reason why I favor new option names is that "allman" and "stroustrup" are the same for function bodies, they're only different on IfStatements and TryStatements. So there's some potential for user confusion on function bodies (where "allman" and "stroustrup" both mean the same thing, but it seems like they shouldn't if we support both options).

**Comment by not-an-aardvark** [comment]
I suggest we change the schema to *add* an optional object, rather than replacing the existing string-based schema.

Current schema:

```json
{
  "rules": {
    "brace-style": ["error", "1tbs"]
  }
}
```

Proposed new schema:

```json
{
  "rules": {
    "brace-style": ["error", "1tbs", {"IfStatement": "allman"}]
  }
}
```

This would allow the user to continue specifying a default brace style, e.g. `1tbs`, and specify exceptions to that default. That way, the user wouldn't have to explicitly list a preference for every possible node type that can have braces in the options object.

> One reason why I favor new option names is that "allman" and "stroustrup" are the same for function bodies, they're only different on IfStatements and TryStatements.

Fair enough. However, I still think it would be better to use the existing names for consistency. That way, users won't have to deal with two separate categories of option strings that largely mean the same thing. IMO, it's acceptable for `{"FunctionDeclaration": "allman"}` and `{"FunctionDeclaration": "stroustrup"}` to enforce the exact same thing.

**Comment by platinumazure** [comment]
I count three total :+1:s (and no :-1:s) from nzakas, kaicataldo, and not-an-aardvark. So although we still need to bikeshed on the schema, I think we can mark the issue as accepted.

@jrsearles What do you think of @not-an-aardvark's [counterproposal](https://github.com/eslint/eslint/issues/6828#issuecomment-263808692)?

**Comment by jsearles** [comment]
My personal preference would be using options like "new-line" and "same-line". I find them more descriptive and discoverable. I don't think developers would typically know specific variants and if they did they would probably use those instead of the individual options.

**Comment by not-an-aardvark** [comment]
> I don't think developers would typically know specific variants and if they did they would probably use those instead of the individual options.

Due to the way the existing options work, a developer would have to know specific variants in order to configure this rule at all. I think it's better to keep the same options for consistency. If we were redesigning the rule from scratch I might agree with you, but I don't think we should have two separate groups of options that mean largely the same thing.

**Comment by jsearles** [comment]
No - it's either/or - at least in the proposal. Either a string of a known variant or an object literal with individual options.

**Comment by not-an-aardvark** [comment]
[My variant](https://github.com/eslint/eslint/issues/6828#issuecomment-263808692) of the proposal suggested using both a string and an object literal.

I don't think having only an object literal is a good idea, because it would require the user to manually list a configuration for every type of node that can have braces. (Regardless, I don't think it's a good idea for the object-literal variant to accept completely different option names from the original.)

Based on the original proposal, it looks like order to distinguish between `1tbs` and `stroustrup` styles, you're introducing two new options, `noCuddledElse` and `noCuddledCatchFinally`. I think these options are unnecessary; we wouldn't need them if we just used `1tbs` and `stroustrup` in the node-types object.

**Comment by jsearles** [comment]
I still think explicit options are more meaningful and discoverable, but i'll leave it to the maintainers to decide.

**Comment by platinumazure** [comment]
Maybe we could consider changing the option names in a semver-major
release, if other users agree that the option names need to be changed. In
order to get this functionality implemented, can we use the old option
names?

**Comment by not-an-aardvark** [comment]
@jsearles Fair enough (and I appreciate that you're expressing your opinion on this, it's definitely useful to have an idea of what users want with the option).

In general, I agree that explicit options are better than implicitly switching behavior. But in this case, I think the existing options are reasonably explicit. (For example, even though not everyone knows what Stroustrup brace style is, the Stroustrup brace style *does* clearly imply specific behavior with `else` statements. It's not like the rule is hazarding a guess). Based on this, I don't think it's worth introducing additional option names when the existing ones are satisfactory.

If we did introduce additional names, I would want to do it in a semver-major change across all of the rule's options, as @platinumazure suggested. But for now I think it would be better to stick with the existing names.

**Comment by platinumazure** [comment]
@jrsearles It's been a while, just wanted to know if you had any thoughts on recent comments. It'd be great to coalesce on something so we can move this forward. Can we maybe just use the old style names (allman/1tbs/stroustrup) for now, and then discuss changing the names as a semver-major change in a separate issue?

**Comment by jrsearles** [comment]
Yeah - sorry. I thought it'd been settled. I'll try to update my PR to reflect these changes this weekend.

**Comment by platinumazure** [comment]
@jrsearles Sorry for losing track of this, but where are we here? Have I missed your PR?

**Comment by craigcosmo** [comment]
@jrsearles Can you make your PR? This is great stuff

**Comment by fatcerberus** [comment]
I really want to see this implemented too, as the current `brace-style` implementation is not nearly granular enough.  My style is mostly Stroustrup, but for class, function and method declarations I use the Allman style with the braces on their own line.  There's no way to express this to ESLint so I'm currently forced to disable the `brace-style` rule entirely.

**Comment by hampustagerud** [comment]
Better control over the curly braces would be awesome since (like @fatcerberus described) the current options are very limiting and some more control would be greatly appreciated. Any updates @jrsearles? 🙂

**Comment by jsearles** [comment]
I haven't had a chance to do anything with this, but I did make a [plugin](https://github.com/jrsearles/eslint-plugin-brace-rules) to support this extensibility.

**Comment by trusktr** [comment]
If you could please update the docs and make sure it works in the latest ESLint, we'd be 🙏 🙌 .

**Comment by nzakas** [comment]
Closing this issue as there is a [plugin](https://github.com/jrsearles/eslint-plugin-brace-rules) that implements the functionality requested in this issue. We :heart: the ESLint community for filling in the gaps!
